MDANSE/Src/MDANSE/Framework/AtomGrouping/grouping.py

```diff
@@ -0,0 +1,348 @@
+            else:
+                iterable = it.product(eles_i, eles_j)
+            labels = [((grp_i, grp_j, *pair), "") for pair in iterable]
+
+            results = (
+                weighted_sum(output_data, result_name + "/<%s><%s>/%s%s", labels) / conc
+            )
+
+            group_id = PAIR_GROUP_TEMPLATE.format(result_name, grp_i, grp_j, post_label)
+
+            output_data.add(
+                group_id,
+                data_type,
+                results.shape,
+                **kwargs,
+            )
+            output_data[group_id][...] = post_func(results)
+            if scaling_factor:
+                output_data[group_id].scaling_factor = conc
+    else:
+        raise NotImplementedError("Grouped total for dim > 2 not implemented.")
+
+
+def label_pairs(labels: Iterable[str], *, all_pairs: bool) -> list[tuple[str, str]]:
+    """
+    Parameters
+    ----------
+    labels : Iterable[str]
+        List of labels.
+    all_pairs : bool
+        Return all pairs if true or only the unique pairs if false.
+
+    Returns
+    -------
+    list[tuple[str, str]]
+        A list of label pairs.
+    """
+    if all_pairs:
+        iterable = it.product(labels, repeat=2)
+    else:
+        iterable = it.combinations_with_replacement(labels, 2)
+    return sorted(iterable)
+
+
+def pair_labels(
+    trajectory: Trajectory, *, intra: bool = False, all_pairs: bool = False
+) -> list[tuple[str, tuple[str, str]]]:
+    """Generates pair labels.
+
+    Parameters
+    ----------
+    trajectory: Trajectory
+        Current state of the trajectory, including selection and transmutation
+    intra : bool
+        Returns the intra label data if true.
+    all_pairs : bool
+        Returns all pairs of labels e.g. OH and HO.
+
+    Returns
+    -------
+    list[tuple[str, tuple[str, str]]]
+        The labels of the results and the labels of the individual
+        atoms in a tuple.
+    """
+    labels = []
+
+    if trajectory._grouping_level == "atom":
+        selected_elements = trajectory.unique_names
+        for ele_i, ele_j in label_pairs(selected_elements, all_pairs=all_pairs):
+            labels.append((f"{ele_i}{ele_j}", (ele_i, ele_j)))
+        return labels
+
+    if intra:
+        for grp in trajectory.group_lookup:
+            eles = sorted(
+                set(
+                    trajectory.atom_types[index]
+                    for cluster in trajectory.chemical_system._clusters[grp]
+                    for index in cluster
+                )
+            )
+            for ele_i, ele_j in label_pairs(eles, all_pairs=all_pairs):
+                pair_label = f"<{grp}>/{ele_i}{ele_j}"
+                label_i = f"<{grp}>/{ele_i}"
+                label_j = f"<{grp}>/{ele_j}"
+                labels.append((pair_label, (label_i, label_j)))
+        return labels
+
+    for grp_i, grp_j in label_pairs(
+        trajectory.group_lookup.keys(), all_pairs=all_pairs
```

Comment:
```suggestion
        trajectory.group_lookup, all_pairs=all_pairs
```
Strictly if this is a dictionary don't need the `keys`